rust/src/mainthread.rs

```diff
@@ -0,0 +1,204 @@
+//! This module provides two ways to execute "jobs":
+//!
+//! 1. On the Binary Ninja main thread (the UI event thread when running in the GUI application):
+//!    * [execute_on_main_thread]
+//!    * [execute_on_main_thread_and_wait]
+//! 2. On a worker thread
+//!
+//! Any manipulation of the GUI should be performed on the main thread, but any
+//! non-GUI work is generally better to be performed using a worker. This is
+//! especially true for any longer-running work, as the user interface will
+//! be unable to update itself while a job is executing on the main thread.
+//!
+//! There are three worker queues, in order of decreasing priority:
+//!
+//!    1. The Interactive Queue ([worker_interactive_enqueue])
+//!    2. The Priority Queue ([worker_priority_enqueue])
+//!    3. The Worker Queue ([worker_enqueue])
+//!
+//! All of these queues are serviced by the same pool of worker threads. The
+//! difference between the queues is basically one of priority: one queue must
+//! be empty of jobs before a worker thread will execute a job from a lower
+//! priority queue.
+//!
+//! The default maximum number of concurrent worker threads is controlled by the
+//! `analysis.limits.workerThreadCount` setting but can be adjusted at runtime via
+//! [set_worker_thread_count].
+//!
+//! The worker threads are native threads, managed by the Binary Ninja core. If
+//! more control over the thread is required, consider using the
+//! [crate::backgroundtask::BackgroundTask] class.
+
+use core::{ffi, mem, ptr};
```

Comment:
Since there's no realistic chance of ever being `#![no_std]`, these should be imported from `std` instead. Also, types like `NonNull`, `CStr`, `c_char`, etc. should probably be imported here as well to avoid extra `::` when using those types (they're used a lot in this file).